### PR TITLE
[MemProf] Use remove_if to erase MapVector elements in bulk

### PR DESCRIPTION
**Pull request by teresajohnson**
A cycle profile showed that we were spending a lot of time invoking
MapVector::erase. According to
https://llvm.org/docs/ProgrammersManual.html#llvm-adt-mapvector-h,
erasing elements one at a time is very inefficient for MapVector and it
is better to use remove_if.

This change resulted in around 7% time reduction on a large thin link.

While here remove an unused function that also invokes erase on MapVectors.

### CODE REVIEW

File: llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp [hunk 1]
```diff
@@ -1694,11 +1682,10 @@ void CallsiteContextGraph<DerivedCCG, FuncTy,
   MapVector<CallInfo, ContextNode *> TailCallToContextNodeMap;
 
   for (auto Entry = NonAllocationCallToContextNodeMap.begin();
-       Entry != NonAllocationCallToContextNodeMap.end();) {
+       Entry != NonAllocationCallToContextNodeMap.end(); Entry++) {
```
Comment (kazutakahirata) [inline llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:1685]: While you are here, may I suggest a range-based for loop here?

**Review by kazutakahirata — APPROVED** [review]
LGTM modulo a minor comment about a range-based for loop.

File: llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp (hunk 1 above)
Comment (teresajohnson) [reply llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:1685]: Yeah, good idea, we can do that now. My local branch seems to have gotten messed up, so I am going to merge this one, then make a separate NFC follow on change to do that.

Comment (teresajohnson) [reply llvm/lib/Transforms/IPO/MemProfContextDisambiguation.cpp:1685]: Done in https://github.com/llvm/llvm-project/pull/94308

**Pull request merged**